File: solutions/csharp/pig-latin/PigLatin.cs
using System;
using System.Text.RegularExpressions;
using System.Linq;

public static class PigLatin
{
    public static string Translate(string sentence) =>
        String.Join(' ', sentence.Split(' ').Select(Convert));

    public static string Convert(string word)
    {
        var basic = Regex.Match(word, @"^[aeiou]|^xr|^yt");
        if (basic.Success)
            return word + "ay";

        var y = Regex.Match(word, @"^([^aiueo]+)(y.*)$");
        if (y.Success)
            return Ay(y);

        var qu = Regex.Match(word, @"^([^aiueo]*qu)(.*)$");
        if (qu.Success)
            return Ay(qu);

        var consonant = Regex.Match(word, @"^([^aiueo]+)(.*)$");
        if (consonant.Success)
            return Ay(consonant);
        
        return word;
    }

    private static string Ay(Match m) =>
        $"{m.Groups[2]}{m.Groups[1]}ay";
}

File: solutions/lua/pig-latin/pig-latin.lua
local function ay(a, b)
  if a then
    return (b or "") .. a .. "ay"
  end
end
local function vowel_xr_yt(word)
  return ay(word:match('^([aeiou].*)') or word:match('^(xr.*)') or word:match('^(yt.*)'))
end
local function y(word)
  return ay(word:match("^([^aiueo]+)(y.*)$"))
end
local function qu(word)
  return ay(word:match("^([^aiueo]*qu)(.*)$"))
end
local function consonant(word)
  return ay(word:match("^([^aiueo]+)(.*)$"))
end
local function convert(word)
  return vowel_xr_yt(word) or y(word) or qu(word) or consonant(word)
end

return function(phrase)
  local words = {}
  for word in phrase:gmatch("%S+") do
    table.insert(words, convert(word))
  end
  return table.concat(words, " ")
end
